File: perl/spicker/spicker1.0/join_spicker_results.pl
#!/usr/bin/perl
use Getopt::Std;
use small_util qw(addAbsPath $codedir);

$nc=5;
$listf="none";
&parse_input();

#store list of headers in memory
if($listf eq "none"){
    @list=`find $root -name "summary.txt"`;
    for $ref (@list){ $ref=substr($ref,-20,7); } #reference to list elements
}
else{  
    open(in,$listf); @list=<in>; close(in); 
}
open(out,">$outf");
print out "#header toprank toprmsd firstrmsd \n";
foreach $l (@list){
    $subdir=substr($l,0,1);
    $header=substr($l,2,5); #print "$subdir $header\n"; exit;
    print "$header\n";
    #compose absolute path of "summary.txt" file
    $f="$root/$subdir/$header/summary.txt"; #print "f=$f\n";
    #output rmsd of first and top centroid. Output rand of top centroid
    open(in,"$f") or next;
    do($l=<in>)until($l=~'#cluster   rmsd');
    chomp($l=<in>);
    @x=split(' ',$l); $tr=$x[0]; $toprmsd=$firstrmsd=$x[1]; $n=1;
    if($firstrmsd=~'nan'){ print "ERROR: firstrmsd=nan for $header\n";exit;}
    chomp($l=<in>);
    while(!($l=~"rmsd of closests") && $n<$nc){#look first nc centroids
	@x=split(' ',$l); 
	$rank=$x[0]; $rmsd=$x[1];
	if($rmsd=~'nan'){ print "ERROR: rmsd=nan for $header\n";exit;}
	if($rmsd<$toprmsd){
	    $toprmsd=$rmsd;
	    $tr=$rank;
	}
	chomp($l=<in>);
	$n++;
    }
    printf out "$header     %2d     %5.2f     %5.2f\n",$tr,$toprmsd,$firstrmsd;
    close(in);
}
close(out);

sub parse_input(){
    my $f=0;
    if(!getopts('l:o:r:n:h')){ &message(); }
    if($opt_l){ $listf=$opt_l;                   }
    if($opt_n){ $nc=$opt_n;                      }
    if($opt_o){ $outf=$opt_o;              $f++; }
    if($opt_r){ $root=addAbsPath($opt_r);  $f++; }
    if($f<2){ &message(); }
    if($opt_h){ &message(); }
}

sub message(){
    system("clear");
    print "Usage: join_spicker_results.pl -l list -r root -o out\n";
    print " Mandatory:\n";
    print " -r root: root directory (root of outputs)\n";
    print " -o out: output file to store results\n";
    print " Optional:\n";
    print " -l list: list of type \"subdir/header\". If not specify, will use instead Unix \"find\" command to find all occurences of file \"summary.txt\"\n";
    print " -n nc : number of top clusters to consider(def=5)\n";
    print "\n";
    exit(0);
}

File: projects/NMR_constraints/join_spicker_results.py
#!/usr/bin/python
import pdb
import sys,os,re
from inputArgs.inputArgs import inpHand
from utilities.small_utilities import chomp
from jobs.job import gen_lf,pastry

ih=inpHand('Usage: join_spicker_results.py',
           ' -a _A_ind root directory ("summary*.txt" files inside, def=curr dir)',
           ' -c __outf output filename  (def=summary.all)',
           ' -d _A_outd output directory (def=curr dir)',
           ' -e __slack permissivity between first and top cluster(0<=slack<1, def=0.0)',
           ' -f _A_listf filter by list of headers',
           ' -g __logname a logfile containing the passed word must be in the same directory as the file summary*.txt',
           ' -i __rco rco-cutof,f, to filter out those summary.txt files with rco-cutoff'
           )
ih.parse(locals(),sys.argv)

#set defaults
if not ind: ind=os.getcwd()             #;print ind
if not outd: outd=os.getcwd()
if not outf: outf=outd+'/summary.all'
else       : outf=outd+'/'+outf
if not slack:slack=1.0
else        : slack=1-float(slack)
listheaders=''
if listf: listheaders=chomp(open(listf,'r').readlines())
if logname:
    p=re.compile(logname)
    #print logname

pt_out=open(outf,'w')
pt_out.write('#header most-populated-centroid best-rmsd-centroid most-dense-centroid\n');
#print 'find '+ind+' -name "summary*.txt"'; sys.exit(1)
for file in chomp(os.popen('find '+ind+' -name "summary*.txt"').readlines()):
    #pdb.set_trace()
    if logname:
        basedir=os.path.dirname(file) #directory where "summary*.txt" resides 
        logfile=os.popen('ls '+basedir+'/*.log').readline()
        if not p.search(logfile): continue #don't process if required logname not present
    pt=open(file,'r')   #;print 'file=',file
    line=pt.readline()  #;print 'line=',line
    #set header
    if re.compile('Target').search(line):
        header=line[8:13] #first line contains the header
        print header,file
    else: header='xxxxx'
    #skip rest if target not in passed list of headers
    if listheaders:
        if header not in listheaders:
            pt.close()
            continue
    #print file
    #print logfile
    p=re.compile('rank size percent density')
    while not p.search(line):
        line=pt.readline()
        #print line

    #find most dense cluster
    line=pt.readline()
    (denserank,size,percent,topdensity,avrmsd)=line.split()
    topdensity=float(topdensity)
    p=re.compile('\s+\d+\s+\d+\s+\d+')
    line=pt.readline()
    while p.match(line):
        (rank,size,percent,density,avrmsd)=line.split()
        density=float(density)  #;print 'density=',density
        if density>topdensity:
            #print density,'>',topdensity
            denserank=rank
            topdensity=density
        line=pt.readline()
    #print 'denserank=',denserank,'topdensity=',topdensity
    p=re.compile('#cluster')
    #find fist line of centroids info
    while not p.search(line): line=pt.readline()
    #find top cluster
    line=pt.readline()
    (firstrank,firstrmsd)=line.split()
    firstrmsd=float(firstrmsd)
    if firstrank==denserank: densermsd=firstrmsd
    toprank=firstrank    
    toprmsd=firstrmsd
    p=re.compile('\s+\d+\s+\d+')    
    line=chomp(pt.readline())
    while p.match(line):
        (rank,rmsd)=line.split()
        rmsd=float(rmsd)
        if rank==denserank: densermsd=rmsd
        if rmsd<firstrmsd*slack and rmsd<toprmsd:
            toprank=rank
            toprmsd=rmsd
        line=chomp(pt.readline())
    pt.close()
    #write output to file    
    line=header+'  1 %5.2f %3s %5.2f %3s %5.2f\n'%(firstrmsd,toprank,toprmsd,denserank,densermsd)
    pt_out.write(line)
sys.exit(0)
